src/types/mod.rs

```diff
@@ -459,31 +457,44 @@ impl Type {
             return Ok(Arc::clone(data));
         }
 
-        // FIXME there is a race condition (which actually predates this version
-        // of the logic) in case the same type is being finalized from multiple
-        // threads at once.
-        if self.occurs_check.fetch_or(true, Ordering::SeqCst) {
-            return Err(Error::OccursCheck);
+        // First, do occurs-check to ensure that we have no infinitely sized types.
+        let mut occurs_check = HashSet::new();
+        for data in bound.verbose_pre_order_iter::<NoSharing>() {
+            if data.is_complete {
+                occurs_check.remove(&(data.node.as_ref() as *const _));
+            } else if data.n_children_yielded == 0
+                && !occurs_check.insert(data.node.as_ref() as *const _)
+            {
+                return Err(Error::OccursCheck);
+            }
         }
```

Comment:
e690bde94cea795e028e1414f0d61d7c30be42c1: I would love to use the TMR for this, but we don't have finalized types yet. I guess we have to do pointer comparisons...

Also, is `occurs_check.remove(...)` an optimization to keep the hash set small?

Reply:
No, it's because types are allowed to be repeated, they just can't be nested. So you need to remove types from the set once you've recursed out of them.